### PR TITLE
nuke graph traversal optimization

### PR DESCRIPTION
**Pull request by jroy1992**
The visited_dict here is keeping track of all the nodes that have been visited. This dict is shared between all the node type items since if a node has already been checked it need not be checked again and again even if other write node have it as a dependency.
After completion of one entire validation cycle, this visited_dict should be reset to its original state i.e all nodes as  not visited.

Couldn't find an appropriate place to perform this resetting, so have kept it under "if item.type == 'file.nuke':". But this might cause an issue if the nuke script item is unchecked by the user. 
One option considered here was to disable the user's ability to toggle off the 'Nuke Script' item. This would mean to tamper with UI just for 'Nuke Script'.
Another option is to enforce it as workflow to not to turn off check for 'Nuke Script' item.

### CODE REVIEW

**Comment by jroy1992** [comment]
Updated the same for hooks/nuke/publish_files.py in sgtk_apps.
I don't have write permission for submodule repos. Could you pls add me?

File: hooks/tk-multi-publish2/nuke/publish_files.py [hunk 1]
```diff
@@ -145,22 +153,46 @@ def _check_for_knob(node, knob):
...
+    #     if not dependency_list:
+    #         return list(set(nodes))
+    #     else:
+    #         depend = self._collect_nodes_in_graph(dependency_list)
+    #         for item in depend:
+    #             nodes.append(item)
+    #         return list(set(nodes))
+
+    def _collect_nodes_in_graph(self, node, file_paths, valid_paths, show_path):
+        if self.visited_dict[node] == 0:
```
Comment (mbrainerd) [inline hooks/tk-multi-publish2/nuke/publish_files.py]: `if node not in self.visited_dict:`

**Comment by jroy1992** [comment]
Yes it is being re-used across different items. If we have say three writetank node items, the same visited_dict will be used by all three. So if a any node is connected to more than one writetank node, it will be visited by only one and others can safely ignore it since it is already visited.

**Comment by mbrainerd** [comment]
Ah, right. I forgot the plugin hook only gets instantiated once across all tasks 👍

**Pull request merged**